### DISCUSSION TITLE
Python有关多线程编程调用计算资源的疑问

### DISCUSSION THREAD

**Discussion by Lodfy-Feng** [physics-data/faq#156] [Q&A] [answered]
我现在需要进行某种循环（随机数运算）直到生成一个符合统计规律的数据，但一个线程跑实在太慢，是否能用多个线程或线程池对这个callback函数进行计算，只要有一个线程得到valid结果，就返回给我这个结果并stop其他线程（除主线程）。不是很清楚实现方法，想请教各位大佬和助教老师

**Comment by Harry-Chen** [comment, marked as answer]
在 Python 中请不要使用多线程，而是使用 `multiprocessing` 提供的多进程支持。你的这个需求比较复杂，可以考虑使用 `apply_async` 方法，一种思路的伪代码如下：

```python
p = Pool()

results = []
for i in range(n):
    processes.append(p.apply_async(your_func, i))

while True:
    for r in results:
        if r.success():
                final_result = r.get()
                break
    else:
        time.sleep(1)

p.terminate()
```

尽管这样确实有一些脏（直接终止其他进程），但 Python 的语言设计导致可能没有更优雅的方案了。

**Comment by Berrysoft** [comment]
如果可能，建议更换采样算法。多次循环直到取出合格的随机数，听起来效率就不高。
